### PR TITLE
feat: update project contact info api

### PR DESCRIPTION
**Pull request by datphamcode295**
#### What's this PR do?

- [x] API update prọect contact info

#### Screenshots (if appropriate)
image

#### Request
```
curl --location --request PUT 'http://localhost:8200/api/v1/projects/8dc3be2e-19a4-4942-8a79-56db391a0b15/contact-info' \
--header 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2OTkzMjExNDIsImlkIjoiMjY1NTgzMmUtZjAwOS00YjczLWE1MzUtNjRjM2EyMmU1NThmIiwiYXZhdGFyIjoiaHR0cHM6Ly9zMy1hcC1zb3V0aGVhc3QtMS5hbWF6b25hd3MuY29tL2ZvcnRyZXNzLWltYWdlcy81MTUzNTc0Njk1NjYzOTU1OTQ0LnBuZyIsImVtYWlsIjoidGhhbmhAZC5mb3VuZGF0aW9uIiwicGVybWlzc2lvbnMiOlsiZW1wbG95ZWVzLnJlYWQiXSwidXNlcl9pbmZvIjpudWxsfQ.GENGPEucSUrILN6tHDKxLMtj0M0REVMUPC7-XhDMpGM' \
--header 'Content-Type: application/json' \
--data-raw '{
    "clientEmail":"fortress@gmai.com",
    "projectEmail":"fortress@d.foundation",
    "accountManagerID":"2655832e-f009-4b73-a535-64c3a22e558f",
    "deliveryManagerID": "ecea9d15-05ba-4a4e-9787-54210e3b98ce"
}'
```

#### Response
```
{
    "data": {
        "clientEmail": "fortress@d.foundation",
        "projectEmail": "fortress@d.foundation",
        "projectHead": [
            {
                "employeeID": "2655832e-f009-4b73-a535-64c3a22e558f",
                "fullName": "Phạm Đức Thành",
                "displayName": "Thanh Pham",
                "avatar": "https://s3-ap-southeast-1.amazonaws.com/fortress-images/5153574695663955944.png",
                "position": "account-manager"
            },
            {
                "employeeID": "2655832e-f009-4b73-a535-64c3a22e558f",
                "fullName": "Phạm Đức Thành",
                "displayName": "Thanh Pham",
                "avatar": "https://s3-ap-southeast-1.amazonaws.com/fortress-images/5153574695663955944.png",
                "position": "delivery-manager"
            }
        ]
    }
}
```

### CODE REVIEW

File: pkg/handler/project/project.go [hunk 1]
```diff
@@ -1030,3 +1030,134 @@ func (h *handler) UpdateGeneralInfo(c *gin.Context) {
...
+	}
+
+	if !exists {
+		l.Error(err, "project not found")
+		c.JSON(http.StatusNotFound, view.CreateResponse[any](nil, nil, ErrProjectNotFound, body, ""))
+		return
+	}
+
+	// Check account manager exists
+	exist, err := h.store.Employee.Exists(h.repo.DB(), body.AccountManagerID.String())
```
Comment (nguyennh4522) [inline pkg/handler/project/project.go:1080]: `exist` should be consistence

Comment (datphamcode295) [reply pkg/handler/project/project.go:1080]: updated

File: pkg/handler/project/project.go [hunk 2]
```diff
@@ -1030,3 +1030,136 @@ func (h *handler) UpdateGeneralInfo(c *gin.Context) {
...
+// @Param Authorization header string true "jwt token"
+// @Param id path string true "Project ID"
+// @Param Body body UpdateContactInfoInput true "Body"
+// @Success 200 {object} view.UpdateProjectContactInfoResponse
+// @Failure 400 {object} view.ErrorResponse
+// @Failure 404 {object} view.ErrorResponse
+// @Failure 500 {object} view.ErrorResponse
+// @Router /projects/{id}/contact-info [put]
+func (h *handler) UpdateContactInfo(c *gin.Context) {
+	projectID := c.Param("id")
```
Comment (nguyennh4522) [inline pkg/handler/project/project.go:1049]: check type of projectID

File: pkg/handler/project/project.go [hunk 3]
```diff
@@ -1030,3 +1030,136 @@ func (h *handler) UpdateGeneralInfo(c *gin.Context) {
...
+	}
+
+	// Update email info
+	rs, err := h.store.Project.UpdateContactInfo(tx.DB(), project.UpdateContactInfoInput{
+		ClientEmail:  body.ClientEmail,
+		ProjectEmail: body.ProjectEmail,
+	}, projectID)
+
+	if err != nil {
+		l.Error(err, "failed to update project information to db")
```
Comment (nguyennh4522) [inline pkg/handler/project/project.go:1159]: just use `db` or `database`, don't use both

File: pkg/handler/project/request.go [hunk 4]
```diff
@@ -232,3 +232,10 @@ func (input DeleteMemberInput) Validate() error {
 
 	return nil
 }
+
+type UpdateContactInfoInput struct {
+	ClientEmail       string     `form:"clientEmail" json:"clientEmail" binding:"email"`
+	ProjectEmail      string     `form:"projectEmail" json:"projectEmail" binding:"email"`
+	AccountManagerID  model.UUID `form:"accountManagerID" json:"accountManagerID" binding:"required"`
+	DeliveryManagerID model.UUID `form:"deliveryManagerID" json:"deliveryManagerID" binding:"required"`
```
Comment (nguyennh4522) [inline pkg/handler/project/request.go:240]: delivery manager is not required

**Comment by nguyennh4522** [comment]
need updates test_seed

**Comment by datphamcode295** [comment]
> need updates test_seed

**Review by nguyennh4522 — APPROVED** [review]

File: pkg/handler/project/project.go [hunk 5]
```diff
@@ -1030,3 +1034,140 @@ func (h *handler) UpdateGeneralInfo(c *gin.Context) {
...
+		l.Error(err, "failed to delete delivery manager of project in db")
+		c.JSON(http.StatusInternalServerError, view.CreateResponse[any](nil, nil, done(err), body, ""))
+		return
+	}
+
+	// Create new one
+	newAccountManager := model.ProjectHead{
+		ProjectID:  model.MustGetUUIDFromString(projectID),
+		EmployeeID: body.AccountManagerID,
+		JoinedDate: time.Now(),
```
Comment (namnhce) [inline pkg/handler/project/project.go:1137]: time now should be defined as a variable. timeNow := time.Now() then use this value for all function

**Review by namnhce — CHANGES_REQUESTED** [review]

File: pkg/handler/project/project.go (hunk 5 above)
Comment (datphamcode295) [reply pkg/handler/project/project.go]: updated

File: pkg/handler/project/project.go [hunk 6]
```diff
@@ -1030,3 +1041,142 @@ func (h *handler) UpdateGeneralInfo(c *gin.Context) {
...
+	timeNow := time.Now()
+
+	if err := h.store.ProjectHead.UpdateLeftDate(tx.DB(), projectID, body.AccountManagerID.String(), model.HeadPositionAccountManager.String(), timeNow); err != nil {
+		l.Error(err, "failed to delete account manager of project in db")
+		c.JSON(http.StatusInternalServerError, view.CreateResponse[any](nil, nil, done(err), body, ""))
+		return
+	}
+
+	if err := h.store.ProjectHead.UpdateLeftDate(tx.DB(), projectID, body.DeliveryManagerID.String(), model.HeadPositionDeliveryManager.String(), timeNow); err != nil {
+		l.Error(err, "failed to delete delivery manager of project in db")
```
Comment (nguyennh4522) [inline pkg/handler/project/project.go:1137]: update error msg

File: pkg/store/project/project.go [hunk 7]
```diff
@@ -100,13 +100,18 @@ func (s *store) One(db *gorm.DB, id string) (*model.Project, error) {
 func (s *store) UpdateGeneralInfo(db *gorm.DB, body UpdateGeneralInfoInput, id string) (*model.Project, error) {
 	project := &model.Project{}
 
-	project.Name = body.Name
-	project.StartDate = body.StartDate
-	project.CountryID = body.CountryID
-
-	return project, db.Table("projects").Where("id = ?", id).Updates(&project).
+	return project, db.Debug().Model(project).Where("id = ?", id).Updates(map[string]interface{}{"name": body.Name, "start_date": body.StartDate, "country_id": body.CountryID}).
```
Comment (nguyennh4522) [inline pkg/store/project/project.go:103]: remove Debug()

File: pkg/handler/project/project.go [hunk 8]
```diff
@@ -612,11 +612,18 @@ func (h *handler) UpdateMember(c *gin.Context) {
...
+			_, err := h.store.ProjectHead.One(tx.DB(), projectID, body.EmployeeID.String(), model.HeadPositionTechnicalLead.String())
+			if err != nil {
+				if errors.Is(err, gorm.ErrRecordNotFound) {
+					err = h.store.ProjectHead.Create(tx.DB(), head)
+					if err != nil {
+						l.Error(err, "failed to create project head")
+						c.JSON(http.StatusInternalServerError, view.CreateResponse[any](nil, nil, done(err), head, ""))
+						return
+					}
+				}
```
Comment (nguyennh4522) [inline pkg/handler/project/project.go:624]: the logic here as same as update project_head, need update old project_head or create new project_head

File: pkg/store/projecthead/project.go [hunk 9]
```diff
@@ -57,3 +59,12 @@ func (s *store) DeleteByProjectIDAndPosition(db *gorm.DB, projectID string, pos
 func (s *store) HardDeleteByPosition(db *gorm.DB, projectID string, employeeID string, position string) error {
 	return db.Unscoped().Where("project_id = ? AND employee_id = ? AND position = ?", projectID, employeeID, position).Delete(&model.ProjectHead{}).Error
 }
+
+func (s *store) UpdateLeftDate(db *gorm.DB, projectID string, employeeID string, position string, timeNow time.Time) error {
+	return db.Debug().Model(&model.ProjectHead{}).Where("project_id = ? AND employee_id = ? AND position = ?", projectID, employeeID, position).Updates(map[string]interface{}{"left_date": timeNow, "deleted_at": timeNow}).Error
```
[the remark this answers is not in the document]
Comment (nguyennh4522) [reply pkg/store/projecthead/project.go:64]: just update left_date, not deleted_at

File: pkg/handler/project/project.go [hunk 10]
```diff
@@ -1030,3 +1060,175 @@ func (h *handler) UpdateGeneralInfo(c *gin.Context) {
...
+	// Update Account Manager
+	timeNow := time.Now()
+	oldHead, err := h.store.ProjectHead.One(tx.DB(), projectID, model.HeadPositionAccountManager)
+
+	newAccountManager := model.ProjectHead{
+		ProjectID:  model.MustGetUUIDFromString(projectID),
+		EmployeeID: body.AccountManagerID,
+		JoinedDate: timeNow,
+		Position:   model.HeadPositionAccountManager,
+	}
```
Comment (namnhce) [inline pkg/handler/project/project.go:1155]: Cái này sao lại chui vô giữa đây

File: pkg/handler/project/project.go [hunk 11]
```diff
@@ -1030,3 +1060,175 @@ func (h *handler) UpdateGeneralInfo(c *gin.Context) {
...
+		}
+
+		err = h.store.ProjectHead.Create(tx.DB(), &newAccountManager)
+		if err != nil {
+			l.Error(err, "failed to create project head")
+			c.JSON(http.StatusInternalServerError, view.CreateResponse[any](nil, nil, done(err), newAccountManager, ""))
+			return
+		}
+	}
+
```
Comment (namnhce) [inline pkg/handler/project/project.go:1183]: Chỗ này cồng kềnh quá em

**Review by namnhce — CHANGES_REQUESTED** [review]

**Review by namnhce — APPROVED** [review]

**Review by nguyennh4522 — APPROVED** [review]

**Pull request merged**